File: personal/game.py
from personal.nncore import network
from personal.utils import basicFunc
from personal.utils import mathFunc
from personal import storeBestMove
import math
import numpy as np

directions = [
	{"row":  0, "col":  1},
	{"row": -1, "col":  1},
	{"row": -1, "col":  0},
	{"row": -1, "col": -1},
	{"row":  0, "col": -1},
	{"row":  1, "col": -1},
	{"row":  1, "col":  0},
	{"row":  1, "col":  1}
]

net = network.Network()


class Game(object):
	def __init__(self, rows, cols, arrangeList=[], nextColor=0):
		self.rows = rows
		self.cols = cols
		self.NONE = 0
		self.BLACK = 1
		self.WHITE = 2
		self.arrange = arrangeList
		self.blackMove = []
		self.whiteMove = []
		self.nextColor = self.BLACK if nextColor == 0 else nextColor
		self.net = net

	def initialize(self):
		upRow = math.floor((self.rows-1)/2)
		leftCol = math.floor((self.cols-1)/2)
		self.arrange = [[0 for col in range(0, self.cols)] for row in range(0, self.rows)]
		self.putPiece(upRow,   leftCol,   self.BLACK)
		self.putPiece(upRow+1, leftCol,   self.WHITE)
		self.putPiece(upRow,   leftCol+1, self.WHITE)
		self.putPiece(upRow+1, leftCol+1, self.BLACK)

	def clearPiece(self, row, col):
		self.arrange[row][col] = self.NONE

	def putPiece(self, row, col, color):
		self.arrange[row][col] = color

	def canPutPiece(self, row, col, color):
		if self.arrange[row][col] != self.NONE:
			return False
		if len(self.getTurnPieceList(row, col, color)) > 0:
			return True
		return False

	def storeMove(self, row, col, color):
		tmpArrangeList = basicFunc.unsharedCopy(self.arrange)
		if color == self.BLACK:
			self.blackMove.append({"arrange": tmpArrangeList, "color": color, "row": row, "col": col})
		else:
			self.whiteMove.append({"arrange": tmpArrangeList, "color": color, "row": row, "col": col})

	def isOutOfRange(self, row, col):
		if row >= self.rows or col >= self.cols or row < 0 or col < 0:
			return True
		else:
			return False

	def getScore(self, color):
		counter = 0
		for row in range(0, self.rows):
			for col in range(0, self.cols):
				if self.arrange[row][col] == color:
					counter += 1
		return counter

	def getTurnPieceList(self, row, col, color):
		turnPieceList = []
		for i in range(0, len(directions)):
			tmpTurnPieceList = self.getTurnPieceForDirect(row, col, color,  directions[i]['row'], directions[i]['col'])
			for j in range(len(tmpTurnPieceList)):
				turnPieceList.append({"row": tmpTurnPieceList[j]['row'], "col": tmpTurnPieceList[j]['col']})
		return turnPieceList

	def getTurnPieceForDirect(self, row, col, color, y, x):
		checkRow, checkCol = row+y, col+x
		if self.isOutOfRange(checkRow, checkCol):
			return []

		if self.arrange[checkRow][checkCol] == color or self.arrange[checkRow][checkCol] == self.NONE:
			return []

		turnPieceList, turnRows, turnCols = [], [], []
		turnRows.append(checkRow)
		turnCols.append(checkCol)
		checkRow += y
		checkCol += x

		while not(self.isOutOfRange(checkRow, checkCol)):
			if self.arrange[checkRow][checkCol] == self.NONE:
				return []
			elif self.arrange[checkRow][checkCol] == color:
				for i in range(0, len(turnRows)):
					turnPieceList.append({"row": turnRows[i], "col": turnCols[i]})
				return turnPieceList
			turnRows.append(checkRow)
			turnCols.append(checkCol)
			checkRow += y
			checkCol += x
		return turnPieceList

	def canPutPieceOnBoard(self, color):
		canPutList = self.getCanPutList(color)
		for canPut in canPutList:
			if canPut == 1:
				return True
		return False

	def getCanPutList(self, color):
		canPutList = []
		for y in range(self.rows):
			for x in range(self.cols):
				if self.canPutPiece(y, x, color):
					canPutList.append(1)
				else:
					canPutList.append(0)
		return canPutList

	def getWinnersData(self):
		if self.getScore(self.BLACK) > self.getScore(self.WHITE):
			return self.blackMove
		else:
			return self.whiteMove

	def goNextTurn(self):
		self.nextColor = self.WHITE if self.nextColor == self.BLACK else self.BLACK
		if self.canPutPieceOnBoard(self.nextColor):
			return True
		else:
			self.nextColor = self.WHITE if self.nextColor == self.BLACK else self.BLACK
			if self.canPutPieceOnBoard(self.nextColor):
				return True
			else:
				return False

	def turnPiece(self, turnPieceList, color):
		for i in range(0, len(turnPieceList)):
			self.putPiece(turnPieceList[i]["row"], turnPieceList[i]["col"], color)

	def returnMoveList(self, row, col):
		u"""Return array which has next move's row and column. (for neural network)
		:param row: row of next move
		:param col: column of next move
		:return: converted array for neural network
		"""
		moveList = []
		for y in range(0, self.rows):
			for x in range(0, self.cols):
				if y == row and x == col:
					moveList.append(float(1))
				else:
					moveList.append(float(0))
		return moveList

	@staticmethod
	def returnNnInputStoreList(rawArray):
		arrangeList = []
		for value in rawArray:
			arrangeList.append(value[0])
		return arrangeList

	def setNextColor(self, nextColor):
		self.nextColor = nextColor
		return True

	def isEnded(self):
		if self.canPutPieceOnBoard(self.WHITE) or self.canPutPieceOnBoard(self.BLACK):
			return False
		else:
			return True

	def getWinnersColor(self):
		if self.getScore(self.BLACK) > self.getScore(self.WHITE):
			return self.BLACK
		elif self.getScore(self.BLACK) < self.getScore(self.WHITE):
			return self.WHITE
		else:
			return self.NONE

	def goNextWithAutoMove(self, nnFlag=False):
		if nnFlag:
			arrangeList = basicFunc.conv64ListToNnInputList(self.arrange, self.nextColor)
			arrangeList = basicFunc.convInput(arrangeList)
			move = self.net.feedforward(np.array(arrangeList))  # move[0][0:63]
			# move = self.net.feedforward(np.array(arrangeList).T)  # move[0][0:63]
			move = mathFunc.softmax(move)
		else:
			move = np.random.rand(1,64) #move[0][0:63]
		index = np.argmax(move[0]*self.getCanPutList(self.nextColor))
		row, col = divmod(index, 8)
		self.goNextWithManualMove(row, col)

	def goNextWithManualMove(self, row, col):
		self.storeMove(row, col, self.nextColor)
		self.putPiece(row, col, self.nextColor)
		turnPieceList = self.getTurnPieceList(row, col, self.nextColor)
		self.turnPiece(turnPieceList, self.nextColor)
		self.goNextTurn()

	def findBestMove(self):
		winRatio, bestRow, bestCol = 0.0, 0, 0
		for index, value in list(enumerate(self.getCanPutList(self.nextColor))):
			if value == 0:  # means the piece cannot be put
				continue
			tmpGame = Game(8, 8, basicFunc.unsharedCopy(self.arrange), self.nextColor)
			row, col = divmod(index, 8)
			tmpGame.goNextWithManualMove(row, col)
			tmpWinRatio = storeBestMove.calcWinRatio(tmpGame.arrange, tmpGame.nextColor, self.nextColor)
			# print("    row:%d, col:%d, WinRatio:%f" % (row, col, tmpWinRatio))
			if tmpWinRatio >= winRatio:
				bestRow, bestCol, winRatio = row, col, tmpWinRatio
		return bestRow, bestCol, winRatio

File: django/personal/static/personal/js/Game.js
var directions = [
    {row: 0, col: 1},
    {row:-1, col: 1},
    {row:-1, col: 0},
    {row:-1, col:-1},
    {row: 0, col:-1},
    {row: 1, col:-1},
    {row: 1, col: 0},
    {row: 1, col: 1}
]
function Game(rows, cols){
    this.rows = rows;
    this.cols = cols;
    this.NONE  = 0;
    this.BLACK = 1;
    this.WHITE = 2;
    this.nextColor = this.BLACK;
    this.blackMove = [];
    this.whiteMove = [];
}

Game.prototype.init = function(){
    var upRow   = Math.floor((this.rows-1)/2);
    var leftCol = Math.floor((this.cols-1)/2);
    this.arrange = [];
    for (var i=0; i<this.rows; i++){
        this.arrange[i] = [];
        for (var j=0; j<this.cols; j++){
            this.clearPiece(i, j);
        }
    }
    this.putPiece(upRow  , leftCol  , this.BLACK);
    this.putPiece(upRow+1, leftCol  , this.WHITE);
    this.putPiece(upRow  , leftCol+1, this.WHITE);
    this.putPiece(upRow+1, leftCol+1, this.BLACK);
}

Game.prototype.clearPiece = function(row, col){
    this.arrange[row][col] = this.NONE;
}

Game.prototype.putPiece = function(row, col, color){    
    this.arrange[row][col] = color;
}

Game.prototype.canPutPiece = function(row, col, color){
    if(this.arrange[row][col]!==this.NONE){
        return false;
    }
    if(this.getTurnPieceList(row, col, color).length>0){
        return true;
    }
    return false;    
}

Game.prototype.storeMove = function(row, col, color){
    var currentArrange = unsharedCopy(this.arrange)
    if(color===this.BLACK){
        this.blackMove.push({arrange:currentArrange, row:row, col:col, color:color});
    }else{
        this.whiteMove.push({arrange:currentArrange, row:row, col:col, color:color});
    }
}

Game.prototype.isOutOfRange = function(row, col){
    if(row>=this.rows || col>=this.cols || row<0 || col<0){
        return true;
    }else{
        return false;
    }
};

Game.prototype.getScore = function(color){
    var counter = 0;
    for(var row=0; row<this.rows; row++){
        for(var col=0; col<this.cols; col++){
            if(this.arrange[row][col]===color){
                counter++;
            }
        }
    }
    return counter;
}

Game.prototype.getTurnPieceList = function(row, col, color){
    var turnPieceList = [];
    for(var i=0; i<directions.length; i++){
        var tmpTurnPieceList = this.getTurnPieceForDirect(row, col, color,  directions[i]['row'], directions[i]['col']);
        if(tmpTurnPieceList.length>0){
            for(var j=0; j<tmpTurnPieceList.length; j++){
                turnPieceList.push({row:tmpTurnPieceList[j]['row'], col:tmpTurnPieceList[j]['col']});
            }
        }
    }
    return turnPieceList;
};

Game.prototype.getTurnPieceForDirect = function(row, col, color, y, x){
    var turnPieceList = [];
    if(this.isOutOfRange(row+y, col+x)){
        return [];
    }
    var checkRow = row+y;
    var checkCol = col+x;
    if(this.arrange[checkRow][checkCol]===color || this.arrange[checkRow][checkCol]===this.NONE){
        return [];
    }
    var turnRows = [];
    var turnCols = [];
    turnRows.push(checkRow);
    turnCols.push(checkCol);
    checkRow += y;
    checkCol += x;
    while(!this.isOutOfRange(checkRow, checkCol)){
        if(this.arrange[checkRow][checkCol]===this.NONE){
            return [];
        }else if(this.arrange[checkRow][checkCol]===color){
            for (var i=0; i<turnRows.length; i++){
                turnPieceList.push({row:turnRows[i], col:turnCols[i]});
            }
            return turnPieceList;
        }
        turnRows.push(checkRow);
        turnCols.push(checkCol);
        checkRow += y;
        checkCol += x;
    }
    return turnPieceList;
}

Game.prototype.canPutPieceOnBoard = function(color){
    canPutList = this.getCanPutList(color);
    for(var i=0; i<canPutList.length; i++){
        if(canPutList[i]===1){
            return true;
        }
    }
    return false;
}

Game.prototype.getCanPutList = function(color){
    var canPutList = [];
    for(y=0; y<this.rows; y++){
        for(x=0; x<this.cols; x++){
            if(this.canPutPiece(y, x, color)){
                canPutList.push(1);
            }else{
                canPutList.push(0);
            }
        }
    }
    return canPutList;
}

Game.prototype.getWinnersData = function(){
    if( this.getScore(this.BLACK) > this.getScore(this.WHITE) ){
        return this.blackMove;
    }else{
        return this.whiteMove;
    }

}

function unsharedCopy(inList){
    returnList = [];
    inList.forEach(function(value){
        returnList.push(value.slice());
    });
    return returnList;
}
